### ISSUE TITLE
Modernization.

### ISSUE THREAD

**Issue by guijan** [resurrecting-open-source-projects/scrot#89]
Hi, I'd like to ask if the scrot project would be open to some changes focused on clean up, code quality and up to date practices. I'll get working on all them as soon as possible if I get a green light.

I apologize for the impromptu wall of text and possibly seeming rash, but I'm a fan of scrot and I can work on improving it in many ways. I'd like to know what the developers think of my ideas before commiting myself to them.

Immediately, I have a few ideas:
* Remove the strdup implementation in [src/utils.c](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/utils.c)

It's a POSIX standard function implemented in all the big free unix systems (musl/glibc/dietlibc Linux, BSDs, Mac) and it [might make it to the upcoming C2x standard](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2353.htm)
scrot depends on imlib2, which uses strdup as ```grep -Ri 'strdup' .``` [in the imlib source tree](https://sourceforge.net/projects/enlightenment/files/imlib2-src/1.7.1/) will attest, and I can't find a place where it implements the function. scrot will be unable to use imlib2 before it's unable to use strdup.
strdup is the only function inside utils.c, so the entire source file and its header can go away.

* Remove the getopt_long implementation in [src/getopt1.c](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/getopt1.c) and [src/getopt.c](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/getopt.c)

If the copyright is any indication, these source files were last updated in 1997 and the only changes since were local fixes for compiler warnings. They're written in pre-ANSI C, and the getopt_long function they implement is provided and actively maintained by all the major free unix systems.
getopt.c is 1047 lines, getopt.c is 187 lines, and getopt.h is 136 lines, which means this useless code is 43% of the scrot source.

* Add a modern build system, and (preferrable but not essential) remove the old one

I'm a fan of [Meson](https://mesonbuild.com/), I will write and maintain Meson build definitions for this project. 
The current build system has a few undesirable properties, both [src/Makefile.am](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/Makefile.am) and [Makefile.am](https://github.com/resurrecting-open-source-projects/scrot/blob/master/Makefile.am) hardcode include directories. [src/Makefile.am](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/Makefile.am) also hardcodes optimization and debug flags (-g -O3).
[configure.ac](https://github.com/resurrecting-open-source-projects/scrot/blob/master/configure.ac) doesn't make use of [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/) to find the libraries the project depends on.

Meson provides a lot of benefits. 
It has integrated test support, so a test suite for scrot could be developed to catch bugs, and Meson also has integrated test coverage reports.
It integrates with IDEs (Visual Studio, XCode, you name it). 
It exclusively builds in a build directory, so the source tree won't be cluttered with object files while working on the project. Meson also automatically triggers rebuilds when you change compile options, and it automatically detects header changes and triggers a rebuild of the affected source files. 
Meson also abstracts away compiler details, so scrot would build unmodified with compilers other than Clang and GCC.
Meson automatically provides some things autotools does manually. Meson automatically figures out what to put inside distfiles, and you only need to ```rm -r``` the build directory to do the equivalent of a ```make clean```.
Meson also has automatic build types, it defaults to debug so while programming the binary will be built with debug symbols and no optimizations. When a package maintainer packages a Meson package, producing a release build with optimization and no debugging symbols is as easy as passing a flag to the build system, same for adding more compiler flags.

* Make use of portable BSD extensions

Some C functions have caveats. strncpy for instance doesn't null terminate when the length of the source string is the same as the length of the destination buffer, leading to a buffer overflow when the string is used. When it does null terminate, it fills the unused space with zeroes pointlessly.
The BSD systems have an alternative called [strlcpy](https://manpages.debian.org/unstable/libbsd-dev/strlcpy.3bsd.en.html) (actually also present on MacOS and the musl libc for Linux) which does not suffer from these issues, and it can be provided by a compatibility library called [libbsd](https://libbsd.freedesktop.org/wiki/). As you can see on the website, the library is widely available among Linux distros, and it isn't needed on BSD and MacOS because they implement the same functions in their libc.
A similar thing is true for strncat, there is an alternative provided by libbsd called [strlcat](https://manpages.debian.org/unstable/libbsd-dev/strlcat.3bsd.en.html).

Additionally, the library is full of useful functions, for instance the singly-linked list in [src/slist.c](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/slist.c) could be provided by the library through [this API](https://manpages.debian.org/unstable/libbsd-dev/queue.3bsd.en.html), moving code out of scrot and into a library.

* Introduce CI

There is a service that provides free CI to FLOSS projects called [Cirrus CI](https://cirrus-ci.org/). With Cirrus, we can build and test scrot on many Linux distributions and also MacOS and FreeBSD automatically. I will admit I don't know very well how CI tests of an X application such as scrot would work, however.

* Rewrite the manual in [mdoc](https://mandoc.bsd.lv/)

mdoc is the "new" format for man pages. It's supported by GNU groff (widely used to provide the man command among Linux distros), mandoc (provides the man command in a few Linux distributions and the BSDs), and even Vim's manual page reader. It can easily be written by a human, and it can also be converted to the classic unix "man" format.
It's better than man because its markup is semantic rather than visual. Because of this the manuals become more consistent and require less work to write, and a reader capable of reading mdoc can search for semantic elements of the manual. For instance, one can search for all manuals with a "-d" flag, or that contain a certain text string in the synopsis.

**Comment by daltomi** [comment]
Hi! @guijan

(I am using a translator, I cannot express myself as much as I want, please excuse me.)

* Remove the strdup implementation
* Remove the getopt_long implementation

I think that two changes that propose is right.

Note: Currently the build system does not link those functions unless he does not find them in the target system. In my case they are always in `strdup@glibc` and `getopt_long@glibc`
I think the file `getopt.c` was included because perhaps, many years ago, some systems did not implement those functions.
For this reason, I left at the PR #86  the local function `strdup` and I forgot to ask the author @languidnights  if it was absolutely necessary for him (I hope not, so we can eliminate it).



* Make use of portable BSD extensions

Yes, I like the idea, I just hope everyone supports `libbsd`

Note: There are 5 calls to the `strncpy` function, 1 to `strcpy`(utils.c delete soon), 3 to `strncat` and 15 to `strcat` IMHO, you can leave `strcat` unchanged as it has enough space (4096 bytes), but that's up to you.



* Add a modern build system,  Introduce CI,  Rewrite the manual in mdoc

I'm not a maintainer so I don't know, the maintainer of this project is @eribertomota  and I want him to have the last decision on these items :-)

Cheers.!!

**Comment by eribertomota** [comment]
Hi guys,

I will check and approve (or not) all document changes. @daltomi, please, check and approve (or not) all programming changes.

I would like to call on you to keep txt2man. I maintain txt2man in Debian and I use it in all projects that I help to maintain in GitHub.

**Comment by eribertomota** [comment]
Also, please, keep autotools.

**Comment by daltomi** [comment]
> #guijan wrote: 
>[src/Makefile.am](https://github.com/resurrecting-open-source-projects/scrot/blob/master/src/Makefile.am) also hardcodes optimization and debug flags (-g -O3).

Yes, on that point I already plan to modify it https://github.com/resurrecting-open-source-projects/scrot/pull/54#discussion_r516239301

> . @daltomi, please, check and approve (or not) all programming changes.

No problem, I'll do the best I can. I hope more C programmers involved in the project. :-)

**Comment by eribertomota** [comment]
> > . @daltomi, please, check and approve (or not) all programming changes.
> 
> No problem, I'll do the best I can. I hope more C programmers involved in the project. :-)

Sure, it is good. @daltomi, Please, remember to poke me when it's time to release a new version.

@guijan, please, always put your name in Copyright in the header of the each changed file.

Thanks guys!

**Comment by guijan** [comment]
Thanks for the responses.


I'll get working on the manual in txt2man format and add my copyright to the meson.build files in my pending PR at #92 later then.

**Comment by eribertomota** [comment]
> Thanks for the responses.
> 
> I'll get working on the manual in txt2man format and add my copyright to the meson.build files in my pending PR at #92 later then.

I'm grateful with your help. However, I would like to ask you again to not migrate from autotools to meson. Please, help us with the main source code.

Cheers,

Eriberto

**Comment by guijan** [comment]
Oh, I understand. I thought adding it without removing autotools was okay.
I've been busy with school tests but I should have the free time to work on scrot by tomorrow.

**Comment by Sherlock4709** [comment]
really interested in where this goes.

**Comment by languidnights** [comment]
> For this reason, I left at the PR #86 the local function `strdup` and I forgot to ask the author @languidnights if it was absolutely necessary for him (I hope not, so we can eliminate it).

My compiler had a standard strdup, so I didn't need it, but it wasn't /officially/ added to the C Standard until 2019 per my reading, so I didn't want to break other people by assuming all compilers would have it.

**Comment by Nable80** [comment]
``man 3 strdup``
> strdup() conforms to SVr4, 4.3BSD, POSIX.1-2001.  strndup() conforms to POSIX.1-2008.

Support of POSIX-incompatible platforms/tools is definitely out of scope here, isn't it?

**Comment by daltomi** [comment]
@languidnights Thanks for answering, do not worry about that, the compilers will support it, if it is not for the standard library it will be for an extension (as long as we don't force the compiler to be strictly C99 which would completely disable POSIX_C_SOURCE compliance)

@Nable80 Agree.

**Comment by guijan** [comment]
Currently, most of the suggestions are complete:
- [x] Remove the strdup implementation in src/utils.c
- [x] Remove the getopt_long implementation in src/getopt1.c and src/getopt.c
- [x] Introduce CI
- [ ] Make use of portable BSD extensions (should wait on #97 because it would require a compatibility library)

Immediately, a few more suggestions are obvious:
- [ ] Write tests (should wait on #97)
- [ ] Polish the English of the txt2man manual

Some were rejected:
* Rewrite the manual in mdoc
* Add a modern build system, and (preferrable but not essential) remove the old one


Right now I don't feel like doing much so the incomplete ones will probably wait a little.

**Comment by guijan** [comment]
I'm considering this PR complete. 
All the initial goals have been met and only the work to implement them is left, and this is clearly stated in the [TODO.md](TODO.md). Further suggestions will be made into their own issues and PRs to avoid snowballing this one with more and more ideas.

**Comment by TimB87** [comment]
> Yes, I like the idea, I just hope everyone supports `libbsd`

I know that CRUX does not update scrot because depending on libbsd is as bad as it can get on a source based distro. It will link to next to everything and leave your system crippled when you remove it (e.g. it leaves bsdtar not working if it was build with libbsd).

Since this pr is done and merged, this comes a bit late, but no, not everybody supports libbsd (even if CRUX does have a port for it, individual maintainers still choose to not use it).

**Comment by thesamesam** [comment]
> > Yes, I like the idea, I just hope everyone supports `libbsd`
> 
> I know that CRUX does not update scrot because depending on libbsd is as bad as it can get on a source based distro. It will link to next to everything and leave your system crippled when you remove it (e.g. it leaves bsdtar not working if it was build with libbsd).
> 
> Since this pr is done and merged, this comes a bit late, but no, not everybody supports libbsd (even if CRUX does have a port for it, individual maintainers still choose to not use it).

But it's optional in Autotools here in scrot? Just pass the configure argument.

And most things will break if you remove a library they're linked against. The issue you're describing sounds like a problem in individual build systems ("automagically dependencies") if stuff uses libbsd simply because it's installed.

**Comment by TimB87** [comment]
> But it's optional in Autotools here in scrot? Just pass the configure argument.

(Fair warning, I didn't dig into it too much, I use and maintain the wayland ports on CRUX and scrot doesn't have a part in that anymore)

With `--without-libbsd` it throws
`configure: error: BSD functions not found and --without-libbsd was used`
 while running configure, then exits.

So my take is that the README.md is not really clear. It depends on libbsd if you are not running on BSD, or whats the deal? 

> And most things will break if you remove a library they're linked against. The issue you're describing sounds like a problem in individual build systems ("automagically dependencies") if stuff uses libbsd simply because it's installed.

True, but if it can be avoided... Not every user digs into the internals of every package they install, so this is really a mean bomb to deploy.

> (Plus libbsd itself doesn't have any dependencies other than its sibling libmd?)

(Also true, but that doesn't make it better?)
 
> Note that this isn't a problem for us at all in Gentoo for the aforementioned reasons even though we _chose_ to enable libbsd here.

Well, I wish I could say the same :) Still, thanks for answering back. I just think its ironic to allow libbsd as a dependency but not migrate to meson or anything other than autotools tbh.

**Comment by guijan** [comment]
I found these new comments completely by chance.  Poking @TimB87.

```markdown
scrot requires a few projects and libraries:
- [libbsd](https://libbsd.freedesktop.org/wiki/) (if `./configure --without-libbsd; make` fails)
```
The libbsd dependency isn't optional. The `./configure --without-libbsd; make` tidbit in the README.md is meant to feature test whether the system implements the libbsd functions we use; if it succeeds, libbsd isn't needed, otherwise, we hard depend on it. 
The BSDs all implement the relevant functions, so does the newlib libc, but glibc, dietlibc, and musl don't.

What Gentoo terms "automagic dependencies" seems to be the issue you're describing under CRUX.  
I added the option because of this [Gentoo article](https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Automagic_dependencies#Why_automagic_dependencies_are_wrong).
Passing --without-libbsd to configure prevents the build system from automagically picking up libbsd from the system and succeeding behind a package maintainer's back, solving the issue.

TL;DR: you should add libbsd to scrot's dependency list to update the CRUX package, it's not optional.

**Comment by TimB87** [comment]
@guijan First off, I am not the package maintainer in CRUX. I am not even a scrot user (happily switched to wayland a long time ago). I am only voicing my personal opinion as well as what I know about the actual package maintainers (as in for the CRUX port) stance.

With that said, CRUX is a glibc based system. So, according to your post, we don't have these features libbsd offers. My personal stance is that I don't want that library anywhere near my systems, and I won't reconsider for just a second. So if I were interested in scrot at all, I would stick to the older version or look for alternatives (that there are).

I am just sad that software like scrot (which I liked for such a long time, so thanks btw) now requires libbsd - that is all. It's your choice of course, I didn't contribute squid to the project so nevermind my ramblings. My personal opinion on libbsd is that it's invasive on a lot of other software, same as icu as an example, so the user always has to make an decision on if they want that or not, as removing the port later on after months of being installed might leave you with a broken system. Big nope, thank you, I'm good.

tl;dr: I don't actually care, BUT if **you** want scrot in CRUX ports database updated, it looks like you'd need to drop the hard dependency on libbsd for glibc based systems.

**Comment by eli-schwartz** [comment]
Wouldn't this generically be a problem solved by ensuring the dependencies of any package that links to libbsd, includes a reference to libbsd?

So it is not the fault of packages that use and need libbsd, that libbsd is on the system (or that it is later removed).

It is, in fact, blaming one project for the *audacity* to use a dependency which *other* projects are known to have "automagic dependencies" on.

So the steps involved here look roughly like:
- Install libbsd because one program (scrot?) *needs* it, and is a reasonable and respectful member of the software ecosystem in the process
- other (misbehaving) programs silently detect its presence, and add magic on it
- remove the first program (scrot?)
- the misbehaving program now fails to run
- Rant at the first program's maintainers, because they are evil and broke the misbehaving program. Clearly, it is not the fault of the misbehaving program that it misbehaves, it is the fault of scrot!
- Profit? $$$


Personally, I would not blame scrot. I would blame the misbehaving program. Then I would either patch the misbehaving program to not use libbsd if the distro decides that that package doesn't need it, or add a dependency on libbsd if the distro decides the package should use it, or, ideally, redirect the rant at the misbehaving program, and also offer an olive branch in the form of a patch that teaches the misbehaving program to take a compile option to select whether to use libbsd.

Then Profit $$$ for real, because the Crux Pkgfile can pass that option to disable it and it will no longer bother anyone.

**Comment by TimB87** [comment]
It's not much of a rant, @eli-schwartz, rather than genuine wonderment, that modernization, as the topic says, rejects a nice meson build system, but instead introduces libbsd as a hard dependency.
As you can read from my earlier comments, the actual scrot maintainer in CRUX rejects libbsd (same as I would do, btw) and already switched to maim. The port will probably be dropped in the near future.

I appreciate your statement but I can't do anything about it. Time is limited and all I wanted to archive here is to maybe make libbsd optional so we could've kept the port. I'm done with it. No hard feelings.

**Comment by eli-schwartz** [comment]
You appreciate my statement, but cannot do anything about it?

Seems like you just don't understand my statement in the first place. Let me try to rephrase it.

The Crux community purportedly has an ethically worrying habit. That habit is called "disapproving of the use of libbsd".

...

This is not about source based distros, binary distros have practically the same problem in transitive build deps. Foo depends on bar, bar is configured to depend on baz, foo optionally supports baz. If the next version of bar removes support for baz, that could break foo... unless the maintainer of foo acts with responsibility and adds a dependency.

The Crux viewpoint is apparently that this means baz is inherently evil software, and any software that dares depend on baz for their own reasons should get responses like "Big nope, thank you, I'm good." and "My distro will now refuse to ever update your software again, what a shame" and "BTW no hard feelings".

Broadly speaking, no linux distribution, either source or binary based, solves this problem. Except Nix of course. Linux distributions that do split development packages (I do not use such a distro myself) have partial, incomplete protection. Linux distributions that automatically scan ELF files for DT_NEEDED tags and add new dependencies, for example, dpkg-shlibdeps, will just find themselves with extra dependencies they didn't expect, and you cannot remove the baz port in the end.

***Automagic dependencies are bad***. I think we all agree on that. But it's not at all classy to lecture a project about how they are "doing it all wrong" as it were, because they *don't* do automagic dependencies, and in fact do everything right, but you are anyways going to characterize the project as a "bad actor" and tell them how much of a bad actor and total troublemaker they are and "hey scrot maintainers, I have genuine wonderment that you would go around causing us so much trouble with this evil work-of-the-devil libbsd thing".

Because you are blaming scrot for the behavior of other projects.

This is problematic.

It is more problematic than all the supposed behaviors of libbsd.

Why.

Are.

You.

Blaming.

This.

Project.

For.

The.

Behavior.

Of.

Other.

Projects.

WHY???

**Comment by eli-schwartz** [comment]
If y'all hate libbsd so much, build it as a static library, then removing the port won't affect anything that builds against it.

**Comment by eli-schwartz** [comment]
TBH, my understanding is that there is no such thing as a project that *optionally* depends on libbsd. The entire point of libbsd is the rationale that the APIs provided by BSD are better, therefore use them.

As such, there are 3 ways that projects will choose to depend on libbsd:
- by looking it up as a system dependency, and requiring that it exist
- by copying extracts of it into their own source tree, and unconditionally building that
- by looking it up as a system dependency, and if that doesn't exist, using copied extracts in their own source tree

This project has done option 1. The problem is, presumably, projects doing option 3.

Given that the alternative to libbsd is using ad-hoc vendored copies of libbsd, there are zero downsides to having libbsd.a available as a static library, and letting projects find that. Either way, they will be linked statically. Only by providing libbsd.a, can you save yourself a bit of work running the compiler on those sources, since they are already built. And per distro policy, any project that could benefit from shared libraries, "isn't supposed to, libbsd is evil, copy it into your source tree instead", so... you would rather statically link it anyway, therefore why not do so via a libbsd port?

As a bonus, not only do you save work by providing precompiled static archives of libbsd-copied-into-your-project, you don't have to get mad at projects which choose option 1 anymore. Everyone wins!

(Obligatory meson comment: meson wraps would allow Crux to build scrot or what-have-you with `--force-fallback-for=libbsd`, which is basically vendoring code excerpts except defaulting to system dependency lookups as a core buildsystem feature, plus you don't have to check other peoples' code into git. Just an ini file with the tarball url and checksum. So I guess there is that.)

**Comment by TimB87** [comment]
Eli, I am at a loss of words!
Yes, I do appreciate you being basically anywhere and spread knowledge.
And still, yes, I can't do anything about it in regards to scrot. My intention wasn't to rain on anyones parade, or rant about "bad development" in scrot, but to voice an opinion to maybe make libbsd optional as in it could work without it.
I can't do squid because it's not my port to care about in the first place.

So.. calm down, please ;)
